File: app/src/main/java/com/example/mainactivity/ui/calendar/CalendarViewModel.kt
package com.example.mainactivity.ui.calendar

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.mainactivity.data.CalendarEventModel
import com.example.mainactivity.data.FamilyRepository
import com.example.mainactivity.data.remote.SupabaseManager
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.filter.FilterOperator
import io.github.jan.supabase.realtime.PostgresAction
import io.github.jan.supabase.realtime.RealtimeChannel
import io.github.jan.supabase.realtime.channel
import io.github.jan.supabase.realtime.postgresChangeFlow
import io.github.jan.supabase.realtime.realtime
import java.time.LocalDate
import java.time.YearMonth
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class CalendarViewModel(app: Application) : AndroidViewModel(app) {
    private val repo = FamilyRepository.get(app)
    private val db get() = SupabaseManager.client.postgrest

    private val _selectedDate = MutableStateFlow(LocalDate.now())
    val selectedDate: StateFlow<LocalDate> = _selectedDate.asStateFlow()

    private val _displayedMonth = MutableStateFlow(YearMonth.now())
    val displayedMonth: StateFlow<YearMonth> = _displayedMonth.asStateFlow()

    private val _events = MutableStateFlow<List<CalendarEventModel>>(emptyList())
    val events: StateFlow<List<CalendarEventModel>> = _events.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private var realtimeChannel: RealtimeChannel? = null
    private var currentUserId: String? = null

    val eventsForSelectedDate: StateFlow<List<CalendarEventModel>> = combine(
        _selectedDate, _events
    ) { date, all ->
        all.filter { event ->
            val from = runCatching { LocalDate.parse(event.dateFrom) }.getOrNull() ?: return@filter false
            val to = runCatching { LocalDate.parse(event.dateTo) }.getOrElse { from }
            !date.isBefore(from) && !date.isAfter(to)
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    init {
        viewModelScope.launch {
            repo.currentUserId.collect { userId ->
                currentUserId = userId
                if (userId != null) loadEvents(userId) else _events.value = emptyList()
            }
        }
        viewModelScope.launch {
            repo.familyChanged.collect {
                val userId = repo.currentUserId.first() ?: return@collect
                loadEvents(userId)
            }
        }
    }

    private suspend fun loadEvents(userId: String) {
        _isLoading.value = true
        runCatching {
            val user = repo.getUser(userId)
            if (user != null) {
                _events.value = if (user.familyId != null) {
                    db.from("calendar_events").select {
                        filter { or { eq("user_id", userId); eq("family_id", user.familyId) } }
                    }.decodeList<CalendarEventModel>()
                        .filter { it.familyId == null || it.familyId == user.familyId }
                } else {
                    db.from("calendar_events").select {
                        filter { eq("user_id", userId) }
                    }.decodeList<CalendarEventModel>()
                        .filter { it.familyId == null }
                }
                if (user.familyId != null) subscribeToEvents(user.familyId, userId)
            }
        }
        _isLoading.value = false
    }

    private suspend fun subscribeToEvents(familyId: String, userId: String) {
        realtimeChannel?.let { runCatching { SupabaseManager.client.realtime.removeChannel(it) } }
        val channel = SupabaseManager.client.channel("calendar-$familyId")
        realtimeChannel = channel
        val flow = channel.postgresChangeFlow<PostgresAction>(schema = "public") {
            table = "calendar_events"
            filter("family_id", FilterOperator.EQ, familyId)
        }
        channel.subscribe()
        viewModelScope.launch { flow.collect { loadEvents(userId) } }
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.launch {
            realtimeChannel?.let { runCatching { SupabaseManager.client.realtime.removeChannel(it) } }
        }
    }

    fun selectDate(date: LocalDate) {
        _selectedDate.value = date
        _displayedMonth.value = YearMonth.of(date.year, date.month)
    }

    fun nextMonth() { _displayedMonth.value = _displayedMonth.value.plusMonths(1) }
    fun prevMonth() { _displayedMonth.value = _displayedMonth.value.minusMonths(1) }

    fun addEvent(
        activity: String,
        allDay: Boolean,
        dateFrom: String,
        dateTo: String,
        timeFrom: String,
        timeTo: String,
        icon: String = "schedule"
    ) = viewModelScope.launch {
        val userId = repo.currentUserId.first() ?: return@launch
        val user = repo.getUser(userId) ?: return@launch
        val resolvedDateTo = if (dateTo.isBlank()) dateFrom else dateTo
        runCatching {
            db.from("calendar_events").insert(buildJsonObject {
                put("user_id", userId)
                if (user.familyId != null) put("family_id", user.familyId)
                put("activity", activity)
                put("all_day", allDay)
                put("date_from", dateFrom)
                put("date_to", resolvedDateTo)
                put("time_from", if (allDay) "" else timeFrom)
                put("time_to", if (allDay) "" else timeTo)
                put("icon", icon)
            })
        }
        loadEvents(userId)
    }

    fun updateEvent(event: CalendarEventModel) = viewModelScope.launch {
        runCatching {
            db.from("calendar_events").update({
                set("activity", event.activity)
                set("all_day", event.allDay)
                set("date_from", event.dateFrom)
                set("date_to", event.dateTo)
                set("time_from", event.timeFrom)
                set("time_to", event.timeTo)
                set("icon", event.icon)
            }) { filter { eq("id", event.id) } }
        }
        val userId = repo.currentUserId.first() ?: return@launch
        loadEvents(userId)
    }

    fun delete(event: CalendarEventModel) = viewModelScope.launch {
        runCatching { db.from("calendar_events").delete { filter { eq("id", event.id) } } }
        val userId = repo.currentUserId.first() ?: return@launch
        loadEvents(userId)
    }
}

File: ios/FamilyApp/Features/Calendar/CalendarViewModel.swift
// Calendar view model.
import Foundation
import Observation
import Supabase

/// java.time.YearMonth equivalent for the month grid.
struct YearMonth: Hashable, Comparable {
    let year: Int
    let month: Int

    static func of(_ date: LocalDate) -> YearMonth {
        YearMonth(year: date.year, month: date.month)
    }

    static func now() -> YearMonth {
        .of(.today())
    }

    var lengthOfMonth: Int {
        LocalDate.daysIn(month: month, year: year)
    }

    func atDay(_ day: Int) -> LocalDate {
        LocalDate(year: year, month: month, day: day)
    }

    func plusMonths(_ count: Int) -> YearMonth {
        let zeroBased = year * 12 + (month - 1) + count
        // Floor toward -inf so month arithmetic wraps correctly for dates before year 0.
        let newYear = zeroBased >= 0 ? zeroBased / 12 : (zeroBased - 11) / 12
        return YearMonth(year: newYear, month: zeroBased - newYear * 12 + 1)
    }

    /// "MMMM yyyy" in the in-app language.
    func formatted(locale: Locale = Locale(identifier: "en_US_POSIX")) -> String {
        let instant = Date(timeIntervalSince1970: TimeInterval(atDay(1).epochDay) * 86400)
        let formatter = DateFormatter()
        formatter.dateFormat = "LLLL yyyy"
        formatter.locale = locale
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter.string(from: instant)
    }

    static func < (lhs: YearMonth, rhs: YearMonth) -> Bool {
        (lhs.year, lhs.month) < (rhs.year, rhs.month)
    }
}

/// The fields for a new calendar event, grouped into a single parameter.
struct EventDraft {
    var activity: String
    var allDay: Bool
    var dateFrom: String
    var dateTo: String
    var timeFrom: String
    var timeTo: String
    var icon = "schedule"
    var isPrivate = false
    var color: Int?
    var attendeeIds: [String] = []
}

@Observable
@MainActor
final class CalendarViewModel {
    private static var cache: [CalendarEventModel] = []

    private(set) var selectedDate = LocalDate.today()
    private(set) var displayedMonth = YearMonth.now()
    private(set) var events: [CalendarEventModel] = CalendarViewModel.cache
    private(set) var familyMembers: [UserModel] = []
    private(set) var isLoading = false
    /// Whether the current user is the family admin — drives creator-or-admin delete gating.
    private(set) var isAdmin = false

    var currentUserId: String? {
        repo.session.currentUserId
    }

    /// Family members other than me — the selectable "Going with" list.
    var otherMembers: [UserModel] {
        familyMembers.filter { $0.id != currentUserId }
    }

    /// The member for an id (event creator / attendee); nil if unknown.
    func member(_ id: String) -> UserModel? {
        familyMembers.first { $0.id == id }
    }

    var eventsForSelectedDate: [CalendarEventModel] {
        events.filter { event in
            guard let from = LocalDate(iso: event.dateFrom) else { return false }
            let to = LocalDate(iso: event.dateTo) ?? from
            return from <= selectedDate && selectedDate <= to
        }
    }

    private let repo: FamilyRepositoryProtocol
    private let observer: RealtimeObserving
    private var subscribedFamilyId: String?
    private var familyChangedTask: Task<Void, Never>?

    init(
        repo: FamilyRepositoryProtocol? = nil,
        realtime: @MainActor () -> RealtimeObserving = { RealtimeObserver() }
    ) {
        self.repo = repo ?? FamilyRepository.shared
        observer = realtime()
        Task { await loadEvents() }
        familyChangedTask = Task { [weak self] in
            guard let stream = self?.repo.familyChanged() else { return }
            for await _ in stream {
                await self?.loadEvents()
            }
        }
    }

    isolated deinit {
        familyChangedTask?.cancel()
    }

    func refresh() {
        Task { await loadEvents() }
    }

    private func loadEvents() async {
        guard let userId = repo.session.currentUserId else {
            events = []
            return
        }
        if events.isEmpty { isLoading = true }
        defer { isLoading = false }
        isAdmin = await repo.isFamilyAdmin(userId: userId)
        guard let user = await repo.getUser(userId) else { return }
        if let familyId = user.familyId {
            familyMembers = await repo.getFamilyMembers(familyId: familyId)
        }

        // Keep the current list on fetch failure rather than clearing it.
        let result = await (try? repo.fetchCalendarEvents(userId: userId, familyId: user.familyId)) ?? events
        Self.cache = result
        events = result

        if let familyId = user.familyId, subscribedFamilyId != familyId {
            subscribedFamilyId = familyId
            observer.start(
                table: "calendar_events",
                scope: familyId,
                filter: .eq("family_id", value: familyId)
            ) { [weak self] in
                await self?.loadEvents()
            }
        }
    }

    // MARK: - Selection

    func selectDate(_ date: LocalDate) {
        selectedDate = date
        displayedMonth = .of(date)
    }

    func nextMonth() {
        displayedMonth = displayedMonth.plusMonths(1)
    }

    func prevMonth() {
        displayedMonth = displayedMonth.plusMonths(-1)
    }

    // MARK: - Mutations

    func addEvent(_ draft: EventDraft) {
        Task {
            guard let userId = repo.session.currentUserId,
                  let user = await repo.getUser(userId) else { return }
            let resolvedDateTo = draft.dateTo.isEmpty ? draft.dateFrom : draft.dateTo

            var temp = CalendarEventModel()
            temp.id = "temp-\(UUID().uuidString)"
            temp.userId = userId
            temp.familyId = user.familyId
            temp.activity = draft.activity
            temp.allDay = draft.allDay
            temp.dateFrom = draft.dateFrom
            temp.dateTo = resolvedDateTo
            temp.timeFrom = draft.allDay ? "" : draft.timeFrom
            temp.timeTo = draft.allDay ? "" : draft.timeTo
            temp.icon = draft.icon
            temp.isPrivate = draft.isPrivate
            temp.color = draft.color
            temp.attendeeIds = draft.attendeeIds
            events.append(temp)

            await repo.insertCalendarEvent(temp)
            await loadEvents()
        }
    }

    func updateEvent(_ event: CalendarEventModel) {
        Task {
            events = events.map { $0.id == event.id ? event : $0 }
            await repo.updateCalendarEvent(event)
            await loadEvents()
        }
    }

    func delete(_ event: CalendarEventModel) {
        Task {
            events.removeAll { $0.id == event.id }
            await repo.deleteCalendarEvent(id: event.id)
            await loadEvents()
        }
    }
}

// MARK: - Pure helpers (unit-tested)

/// Month grid cells: leading nils to Monday-align, then each day, padded to full weeks.
func monthCells(_ month: YearMonth) -> [LocalDate?] {
    let first = month.atDay(1)
    // ISO weekday: Monday=1 … Sunday=7 → offset 0…6. epochDay 0 (1970-01-01) was a Thursday.
    let weekday = ((first.epochDay + 3) % 7 + 7) % 7 + 1
    var cells: [LocalDate?] = Array(repeating: nil, count: weekday - 1)
    for day in 1...month.lengthOfMonth {
        cells.append(month.atDay(day))
    }
    while !cells.count.isMultiple(of: 7) {
        cells.append(nil)
    }
    return cells
}

/// Concise time label for the event card.
func eventTimeLabel(
    _ event: CalendarEventModel,
    locale: Locale = Locale(identifier: "en_US_POSIX")
) -> String {
    if event.allDay { return L("All day", locale: locale) }
    return [event.timeFrom, event.timeTo]
        .filter { !$0.isEmpty }
        .joined(separator: " – ")
}

/// "EEEE, d MMMM" in the in-app language.
func sectionDateLabel(
    _ date: LocalDate,
    locale: Locale = Locale(identifier: "en_US_POSIX")
) -> String {
    let instant = Date(timeIntervalSince1970: TimeInterval(date.epochDay) * 86400)
    let formatter = DateFormatter()
    formatter.dateFormat = "EEEE, d MMMM"
    formatter.locale = locale
    formatter.timeZone = TimeZone(identifier: "UTC")
    return formatter.string(from: instant)
}
